MNT: Cleanup infs handling in np.testing assertion utilities

danra opened the pull request:
Besides just being a nice cleanup, a specific motivation is towards reporting max errors indices (#28827). Without this change, `assert_array_almost_equal.compare` can produce an output shape different from its input, requiring additional logic to produce the correct indices in the original array.

Note: This PR is based over #29318 commits because it somewhat depends on it. It's probably better to review #29318 before this one.

danra committed:
MNT: Remove assert_array_almost_equal old infs logic, but tests fail

A nice cleanup (newer, similar inf handling already exists now in
assert_array_compare), and resolves the shape mismatch issue. However,
the removed logic was handling complex infs while the one isn't, causing
the new test and an existing one (TestInterp::test_complex_interp) to
now fail with RuntimeWarnings attempting to subtract the complex infs.

danra committed:
MNT: assert_array_compare handles all inf values

assert_array_compare now tests all inf values for matching position
and value, including complex infs.

Fixes the failing tests.

danra committed:
TST: Test array_allclose behavior for inf values

The behavior for real infs is the same is before.

For complex infs, demonstrates that the behavior for mismatching
values is now cleaner, showing a concise error message vs.
previously displaying nan max errors.

For complex infs with matching values, the behavior is the same as
before, accepting them as equal (although internally they would
now be filtered ahead of being passed to isclose, like real infs
already had been).

danra committed:
TST: assert_allclose behavior with nans that are also infs

ngoldbaum commented:
Can you elaborate a little bit more about what is broken in the current inf handling and what this PR does exactly to fix it? It's not clear to me just looking at the diff here.

danra committed:
TST: Add failing test showing shape mismatch issue

Recently, #29112 added showing first mismatches indices, but
assert_array_almost_equal.compare trips it up by returning a shape
different from its input, causing an IndexError:

```
    <...>
    if invalids.ndim != 0:
        if flagged.ndim > 0:
>           positions = np.argwhere(np.asarray(~flagged))[invalids]
E           IndexError: boolean index did not match indexed array along axis 0; size of axis is 3 but size of corresponding boolean axis is 2
```
(traceback shown using pytest --full-trace)

danra commented:
Rebased and ready for review. In addition to the #28827 original motivator, since the original PR #29112 was also merged,  also tripping up on the shape mismatch caused by `assert_array_almost_equal`, and also fixed by this PR. Added a commit demonstrating the issue.

danra commented:
> is it possible for this change to cause tests that previously passed to start failing with a numpy version that includes this change?

No.

>  Or similarly, for a test that used to fail to start passing? If it's the latter, was the failure that happened before actually a bug?

The [first commit](https://github.com/numpy/numpy/pull/29321/commits/8ce11317009b886dd18a698b6d6313994c31e567) in the PR demonstrates a case of `assert_array_almost_equal` raising an `IndexError` due to the recently merged #29112. This PR fixes the bug so that `assert_array_almost_equal` can return a real result, whether pass or fail.

Other than that, there should be no impact on `np.testing` assertions pass/fail results. The change is internal in nature to avoid modifying the array shape, towards reporting indexes in #28827. Since the recently merged #29112 also reports indexes, it also needs this fix to work correctly in all cases.

ngoldbaum reviewed (approved):
LGTM! After reading through this commit-by-commit this makes much more sense.

@mhvk would you mind looking at this one as well?

mhvk reviewed (approved):
Thanks, this looks great!  I have two nitpicky comments, which barely are worth rerunning CI for, so feel free to ignore.

I also checked that the astropy test suite passes with this PR.

ngoldbaum commented:
> I have two nitpicky comments, which barely are worth rerunning CI for, so feel free to ignore

You can skip subsets of the CI too, if you don't want everything to run: https://numpy.org/doc/stable/dev/development_workflow.html#commands-to-skip-continuous-integration

danra committed:
MNT: Extract robust_any_difference() helper to DRY

danra commented on numpy/testing/_private/utils.py:
```diff
@@ -804,6 +804,33 @@ def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
+        else:
+            assert infs_mask.all()
+
+        # For details on the work-arounds employed here, see func_assert_same_pos above
```
Extracted `robust_any_difference` for the significant part of the logic

danra commented on numpy/testing/tests/test_utils.py:
```diff
@@ -1336,6 +1358,33 @@ def test_strict(self):
+        assert_allclose(a, b)
+
+        b = np.array([3.])
+        expected_msg = 'inf location mismatch:'
```
I think it's a bit nicer to keep it the same pattern as everywhere else in the file
Also, does a colon not require escaping in a regex? Who knows? :)

danra commented:
The CI failure looks real but unrelated to this change

mhvk commented:
Thanks! The refactoring made this (even) nicer. Let's get it in.
